### PR TITLE
Retry until exception type

### PR DESCRIPTION
**Pull request by Brian-Williams**
Add complement for retry_if_exception_type, which performs the inverse operation of ignoring all results and exceptions until an exception is hit.

Feel free to change name of the class. Also, I understand if this is superfluous and is closed. I needed it in a very specific use case and it may not be useful to others.

### CODE REVIEW

File: tenacity/retry.py [hunk 1]
```diff
@@ -64,14 +64,26 @@ def __call__(self, attempt):
...
 
+class retry_until_exception_type(retry_if_exception):
+    """Retries until an exception is raised of one or more types."""
+
+    def __init__(self, exception_types=Exception):
+        self.exception_types = exception_types
+        super(retry_until_exception_type, self).__init__(
+            lambda e: not isinstance(e, exception_types))
+
+    def __call__(self, attempt):
```
Comment (jd) [inline tenacity/retry.py]: Why do you need to override that?

File: tenacity/tests/test_tenacity.py [hunk 2]
```diff
@@ -382,7 +382,7 @@ def go(self):
 
 
 class NoIOErrorAfterCount(object):
-    "Holds counter state for invoking a method several times in a row."
+    """Holds counter state for invoking a method several times in a row."""
```
Comment (jd) [inline tenacity/tests/test_tenacity.py]: Would be great to have that in a different commit/PR.

File: tenacity/tests/test_tenacity.py [hunk 3]
```diff
@@ -409,14 +409,50 @@ def __init__(self, count):
     def go(self):
         """Raise a NameError until after count threshold has been crossed.
 
-        Tthen return True.
+        Then return True.
```
Comment (jd) [inline tenacity/tests/test_tenacity.py]: ditto.

File: tenacity/retry.py (hunk 1 above)
Comment (Brian-Williams) [inline tenacity/retry.py]: I want to retry even if no exception is raised.

I'm not sure why `retry_if_exception` needs the conditional. I would've simply removed it there, but didn't want to alter code I didn't dive deeply into. Is it an optimization?

Comment (Brian-Williams) [inline tenacity/retry.py]: If it is an optimization a similar version for this class would be:

```
def __call__(self, attempt):
    # always retry if no exception was raised
    if not attempt.failed:
        return True
    return self.predicate(attempt.exception())
```

File: tenacity/tests/test_tenacity.py (hunk 2 above)
Comment (Brian-Williams) [inline tenacity/tests/test_tenacity.py]: I normally would separate the unique changes like I did in #80, but I don't like making style only pull requests as it simply makes git blame harder to read.

Those style changes make sense in that commit IMO. Since the docstrings were copied in the complement functions, we made their style match. And I didn't alter anything I didn't directly copy/change. For example there is still non-triple quoted docstrings for `NoneReturnUntilAfterCount`.

If you prefer style only pull requests I can split it off. If I wasn't adding `IOErrorUntilCount` with a matching docstring to this function I would've never changed this.

File: tenacity/tests/test_tenacity.py (hunk 3 above)
Comment (Brian-Williams) [inline tenacity/tests/test_tenacity.py]: This shouldn't be in this pull request. Apologies.

**Comment by Brian-Williams** [comment]
Thanks for the quick response, jd! :D

I rebased to 2 commits, one adding the functionality and one testing it.

File: tenacity/retry.py [hunk 4]
```diff
@@ -64,14 +64,26 @@ def __call__(self, attempt):
...
-    """Retries if an exception has been raised of a certain type."""
+    """Retries if an exception has been raised of one or more types."""
 
     def __init__(self, exception_types=Exception):
         self.exception_types = exception_types
         super(retry_if_exception_type, self).__init__(
             lambda e: isinstance(e, exception_types))
 
 
+class retry_until_exception_type(retry_if_exception):
```
Comment (jd) [inline tenacity/retry.py]: It's generally admitted that
* `if not` == `unless`
* `while not` == `until`

Since  it's basically the opposite of `retry_if_exception_type` I'd suggest renaming to `retry_unless_exception_type`.

File: tenacity/retry.py (hunk 1 above)
Comment (jd) [inline tenacity/retry.py]: I'd like this optimization. I think it's a good call to not run the predication function, that could be heavy, if no need to.

File: tenacity/tests/test_tenacity.py (hunk 2 above)
Comment (jd) [inline tenacity/tests/test_tenacity.py]: > but I don't like making style only pull requests as it simply makes git blame harder to read.

What makes git blame hard to read is the change itself, not that it is a specific PR.

But it's not big deal. I wrote "would be great" so I can pinch my nose ;)

File: tenacity/tests/test_tenacity.py (hunk 3 above)
Comment (jd) [inline tenacity/tests/test_tenacity.py]: Don't worry, it's no big deal.

File: tenacity/retry.py (hunk 4 above)
Comment (Brian-Williams) [inline tenacity/retry.py]: Makes sense to me. Hardest thing in programming for me is naming stuff.

**Comment by Brian-Williams** [comment]
Renamed, optimized and squashed into one commit with bullet points. AFAIK this is good to merge. Feel free to ping for any other needed changes.
